### PR TITLE
[SPARK-42482][CONNECT] Scala Client Write API V1

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Implemented the basic Dataset#write API to allow users to write the df into tables, csv etc. files.

### Why are the changes needed?
Basic write operation.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Integration tests.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/java/org/apache/spark/sql/SaveMode.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.apache.spark.annotation.Stable;
+
+/**
+ * SaveMode is used to specify the expected behavior of saving a DataFrame to a data source.
+ *
+ * @since 1.3.0
+ */
+@Stable
+public enum SaveMode {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/java/org/apache/spark/sql/SaveMode.java:27]: We could also move the actual class to Catalyst.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 2]
```diff
@@ -128,6 +128,15 @@ class ClientE2ETestSuite extends RemoteSparkSession {
     }
   }
 
+  test("write table") {
+    val df = spark.range(10).limit(3)
+    df.write.mode(SaveMode.Overwrite).saveAsTable("myTable")
+    val list = spark.sql("select * from myTable").collectAsList()
+    assert(list.size() == 3)
+
+    spark.sql("drop table if exists myTable")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:137]: this is lazy. you need to call collect()

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+   *   // Java:
+   *   df.write().text("/path/to/output")
+   * }}}
+   * The text files will be encoded as UTF-8.
+   *
+   * You can find the text-specific options for writing text files in
+   * <a href="https://spark.apache.org/docs/latest/sql-data-sources-text.html#data-source-option">
+   *   Data Source Option</a> in the version you use.
+   *
+   * @since 1.6.0
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:425]: Update annotation?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+/**
+ * Interface used to write a [[Dataset]] to external storage systems (e.g. file systems,
+ * key-value stores, etc). Use `Dataset.write` to access this.
+ *
+ * @since 3.4.0
+ */
+@Stable
+final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+
+  private val df = ds.toDF()
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:37]: Do we need this conversion? That seems like an artifact from the original implementation.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+  }
+
+  private def executeWriteOperation(f: proto.WriteOperation.Builder => Unit): Unit = {
+    val builder = proto.WriteOperation.newBuilder()
+
+    builder.setInput(df.plan.getRoot)
+
+    // Set path or table
+    f(builder)
+    require(builder.hasPath ^ builder.hasTable) // Only one can be set
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:241]: pfff... this is clever. `builder.hasPath != builder.hasTable` might be easier to parse.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+      case SaveMode.Overwrite => proto.WriteOperation.SaveMode.SAVE_MODE_OVERWRITE
+      case SaveMode.Ignore => proto.WriteOperation.SaveMode.SAVE_MODE_IGNORE
+      case SaveMode.ErrorIfExists => proto.WriteOperation.SaveMode.SAVE_MODE_ERROR_IF_EXISTS
+    })
+
+    source.foreach(builder.setSource)
+    sortColumnNames.foreach(names => builder.addAllSortColumnNames(names.asJava))
+    partitioningColumns.foreach(cols => builder.addAllPartitioningColumns(cols.asJava))
+
+    numBuckets.foreach(n => {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:254]: Should we validate somewhere that bucket columns is not empty. We might also want to check that numBuckets > 0?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+   * {{{
+   *   format("orc").save(path)
+   * }}}
+   *
+   * ORC-specific option(s) for writing ORC files can be found in
+   * <a href=
+   *   "https://spark.apache.org/docs/latest/sql-data-sources-orc.html#data-source-option">
+   *   Data Source Option</a> in the version you use.
+   *
+   * @since 1.5.0
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:402]: Update annotation

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 8]
```diff
@@ -2214,6 +2214,18 @@ class Dataset[T] private[sql] (val session: SparkSession, private[sql] val plan:
...
 
+
+  /**
+   * Interface for saving the content of the non-streaming Dataset out into external storage.
+   *
+   * @group basic
+   * @since 3.4.0
+   */
+  def write: DataFrameWriter[T] = {
+    // if (isStreaming) {} TODO: support client local check?
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2225]: Nah that is a server side problem.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 9]
```diff
@@ -128,6 +128,15 @@ class ClientE2ETestSuite extends RemoteSparkSession {
     }
   }
 
+  test("write table") {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:131]: We need more test coverage.

File: connector/connect/client/jvm/src/main/java/org/apache/spark/sql/SaveMode.java (hunk 1 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/main/java/org/apache/spark/sql/SaveMode.java:27]: I checked the MiMaExcludes file and find if we want to break spark-sql back compatibility, we need to have a very good reason to do so. So I leave this out of this PR.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Pull request closed without merging**